### PR TITLE
feat: introduce JwtClaimsSource

### PR DESCRIPTION
**Pull request by platosha**
Fixes vaadin/fusion#46

### CODE REVIEW

File: vaadin-spring-tests/test-spring-security-fusion/src/main/java/com/vaadin/flow/spring/fusionsecurity/SecurityConfig.java
Comment (haijian-vaadin) [inline vaadin-spring-tests/test-spring-security-fusion/src/main/java/com/vaadin/flow/spring/fusionsecurity/SecurityConfig.java:61]: could remove this commented code?

**Review by haijian-vaadin — COMMENTED** [review]
I got a feeling we could remove the extra layer of JwtClaminsSource? instead could use a `JwtClaimAccessor`?

File: vaadin-spring-tests/test-spring-security-fusion/src/main/java/com/vaadin/flow/spring/fusionsecurity/SecurityConfig.java
Comment (platosha) [reply vaadin-spring-tests/test-spring-security-fusion/src/main/java/com/vaadin/flow/spring/fusionsecurity/SecurityConfig.java:61]: Temporarily commented for testing. Maybe I should rather put it back, we probably want to keep `setStatelessAuthentication` method tested.

**Comment by platosha** [comment]
> I got a feeling we could remove the extra layer of `JwtClaimsSource`? instead could use a `JwtClaimAccessor`?

Not sure if it’s nice in practice. We need a converter from the authentication data to the JWT claims set.  `JwtClaimAccessor` basically is meant to represent the JWT claims set itself, so it’s not usable as a converter.

One way about it is to support a generic `Function<Authentication, JwtClaimAccessor>` type of converter. I have an impression that use case specific types are a better fit though, especially for dependency injection that is type-dependant.

**Comment by platosha** [comment]
My plan here is to replace `JwtClaimAccessor` with even more specific [`JwtClaimsSet`](https://github.com/spring-projects/spring-security/blob/main/oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtClaimsSet.java) now that it is released.

File: vaadin-spring/src/main/java/com/vaadin/flow/spring/security/stateless/JwtSecurityContextRepository.java
Comment (haijian-vaadin) [inline vaadin-spring/src/main/java/com/vaadin/flow/spring/security/stateless/JwtSecurityContextRepository.java:142]:
```suggestion
        JwtClaimAccessor claimAccessor = jwtClaimsSource.get(authentication);
```

File: vaadin-spring/src/main/java/com/vaadin/flow/spring/security/stateless/DefaultJwtClaimsSource.java
Comment (haijian-vaadin) [inline vaadin-spring/src/main/java/com/vaadin/flow/spring/security/stateless/DefaultJwtClaimsSource.java:31]: I think it's better to make it not final, so that as a developer, I can create my own JwtClaimSource by extending the default one

**Review by haijian-vaadin — CHANGES_REQUESTED** [review]

File: vaadin-spring/src/main/java/com/vaadin/flow/spring/security/stateless/DefaultJwtClaimsSource.java
Comment (platosha) [reply vaadin-spring/src/main/java/com/vaadin/flow/spring/security/stateless/DefaultJwtClaimsSource.java:31]:
```suggestion
public class DefaultJwtClaimsSource implements JwtClaimsSource {
```

**Review by haijian-vaadin — APPROVED** [review]

**Review by haijian-vaadin — APPROVED** [review]

**Review by haijian-vaadin — CHANGES_REQUESTED** [review]
Per discussion, let's wait until #940 and #944 are merged, then we use `JwtClaimSet` instead `JwtClaimAccessor`

**Comment by Artur-** [comment]
This should go into https://github.com/vaadin/flow

**Pull request closed without merging**
